// passport-mfp-token-validation/lib/analytics-sdk-node/constants.js
/**
 * @license
 * Licensed Materials - Property of IBM
 * 5725-I43 (C) Copyright IBM Corp. 2006, 2013. All Rights Reserved.
 * US Government Users Restricted Rights - Use, duplication or
 * disclosure restricted by GSA ADP Schedule Contract with IBM Corp.
 */

/*global module*/

// NOTE: If you add any new constants you also need to update the analytics-sdk-node/test/unit/constants.spec.js unit test.
module.exports = {
    MobileServerNetworkTransactions: 'MobileServerNetworkTransactions',

    globalTrackingId: 'globalTrackingId',
    inboundRequestURL: 'inboundRequestURL',
    inboundTimestamp: 'inboundTimestamp',
    outboundTimestamp: 'outboundTimestamp',
    outboundRequestURL: 'outboundRequestURL',
    outboundBackendTimestamp: 'outboundBackendTimestamp',
    requestProcessingTime: 'requestProcessingTime',
    inboundBackendTimestamp: 'inboundBackendTimestamp',
    responseProcessingTime: 'responseProcessingTime',
    duration: 'duration',

    iss: 'iss',
    sub: 'sub',
    aud: 'aud',
    exp: 'exp',
    iat: 'iat',
    userId: 'userId',
    userAuthBy: 'userAuthBy',
    userDisplayName: 'userDisplayName',
    userAttributesRole: 'userAttributesRole',
    deviceId: 'deviceId',
    devicePlatform: 'devicePlatform',
    deviceOsVersion: 'deviceOsVersion',
    deviceModel: 'deviceModel',
    appId: 'appId',
    appVersionNumber: 'appVersionNumber',

    tokenIss: 'iss',
    tokenSub: 'sub',
    tokenAud: 'aud',
    tokenExp: 'exp',
    tokenIat: 'iat',
    tokenImfUser: 'imf.user',
    tokenUserId: 'id',
    tokenUserAuthBy: 'authBy',
    tokenUserDisplayName: 'displayName',
    tokenUserAttributes: 'attributes',
    tokenUserAttributesRole: 'role',
    tokenImfDevice: 'imf.device',
    tokenDeviceId: 'id',
    tokenDevicePlatform: 'platform',
    tokenDeviceOsVersion: 'osVersion',
    tokenDeviceModel: 'model',
    tokenImfApplication: 'imf.application',
    tokenImfApplicationId: 'id',
    tokenImfApplicationVersion: 'version',

    //Environments
    bluemix: 'bluemix',
    onpremise: 'onpremise',
    localMode: 'local-mode',

    // Analytics Keys
    keySubscribeTags: 'subscribeTags',
    keyUnsubscribeTags: 'unsubscribeTags',
    keyAppID: 'appID',
    keyAppVersionCode: 'appVersionCode',
    keyTimestamp: 'timestamp',
    keyTimezone: 'timezone',
    keyCreatorUserID: 'creatorUserID',
    keyTitle: 'title',
    keyFlowDestination: 'flowDestination',
    keyFlowSource: 'flowSource',
    keyMetrics: 'metrics',
    keyChartProperties: 'chartProperties',
    keyCustomEventType: 'customEventType',
    keyCustomChartType: 'customChartType',
    keyCustomXAxis: 'customXAxis',
    keyCustomYAxis: 'customYAxis',
    keyTerm: 'term',
    keyFilters: 'filters',
    keyTableColumns: 'tableColumns',
    keyDeviceID: 'deviceID',
    keyDeviceModel: 'deviceModel',
    keyDeviceBrand: 'deviceBrand',
    keyDeviceOS: 'deviceOS',
    keyDeviceOSVersion: 'deviceOSversion',
    keyAppName: 'appName',
    keyAppVersion: 'appVersion',
    keyServerIpAddress: 'serverIpAddress',
    keyServerIpAddresses: 'serverIpAddresses',
    keyHostname: 'hostname',
    keyWifiAPS: 'wifiAPS',
    keyWifiConnectedSSID: 'wifiConnectedSSID',
    keyWifiConnectedMAC: 'wifiConnectedMAC',
    keyWifiTimestamp: 'wifiTimestamp',
    keyLevel: 'level',
    keyPkg: 'pkg',
    keyMessage: 'message',
    keyStacktrace: 'stacktrace',
    keyStacktraceMessage: 'stacktraceMessage',
    keyMethod: 'method',
    keyFile: 'file',
    keyLine: 'line',
    keyTrackingID: 'trackingID',
    keyResourceURL: 'resourceURL',
    keyUserAgent: 'userAgent',
    keyDeviceIpAddress: 'deviceIpAddress',
    keyCategory: 'category',
    keyNetworkType: 'networkType',
    keyRoundTripTime: 'roundTripTime',
    keyRoundTripTimeCount : 'roundTripTimeCount',
    keyStartTimestamp: 'startTimestamp',
    keyEndTimestamp: 'endTimestamp',
    keyBytesSent: 'bytesSent',
    keyBytesSentCount: 'bytesSentCount',
    keyBytesReceived: 'bytesReceived',
    keyBytesReceivedCount: 'bytesReceivedCount',
    keyNotificationID: 'notificationID',
    keyNotificationAction: 'notificationAction',
    keyFirstAccessTimestamp: 'firstAccessTimestamp',
    keyLastAccessTimestamp: 'lastAccessTimestamp',
    keyUniqueIDs : 'uniqueIDs',
    keyUserAuthBy: 'userAuthBy',
    keyUserDisplayName: 'userDisplayName',
    keySessionID: 'sessionID',
    keySourceMethodName: 'sourceMethodName',
    keyLoggerName: 'loggerName',
    keySource: 'source',
    keyThreadID: 'threadID',
    keySourceClass: 'sourceClass',
    keyOutboundRequestURL: 'outboundRequestURL',
    keyInboundTimestamp: 'inboundTimestamp',
    keyOutboundTimestamp: 'outboundTimestamp',
    keyServerProcessingTime: 'serverProcessingTime',
    keyServerProcessingTimeCount: 'serverProcessingTimeCount',
    keyBackendProcessingTime: 'backendProcessingTime',
    keyBackendProcessingTimeCount: 'backendProcessingTimeCount',
    keyCompressed: 'compressed',
    keyServiceName: 'serviceName',
    keyAdapterName: 'adapterName',
    keyProcedureName: 'procedureName',
    keyTag: 'tag',
    keyNotificationCountIOS: 'notificationCountIOS',
    keyNotificationCountAndroid: 'notificationCountAndroid',
    keyNotificationCountWindows: 'notificationCountWindows',
    keyNotificationCountWindowsPhone8: 'notificationCountWindowsPhone8',
    keyMessagePushTimestamp: 'messagePushTimestamp',
    keyMessageText: 'messageText',
    keyProduction: 'production',
    keyDestNumber: 'destNumber',
    keySenderID: 'senderID',
    keyEventSource: 'eventSource',
    keySubscriptionProperties: 'subscriptionProperties',
    keyMediator: 'mediator',
    keyMessageCount: 'messageCount',
    keyAuthSuccess: 'authSuccess',
    keyValidationCode: 'validationCode',
    keyRealmName: 'realmName',
    keyLoginModule: 'loginModule',
    keyLoginModuleName: 'loginModuleName',
    keyAuthenticator: 'authenticator',
    keyResponseCode: 'responseCode',
    keyApps: 'apps',
    keyDeviceApps: 'deviceApps',
    keyDuration: 'duration',
    keyDurationCount : 'durationCount',
    keyClosedBy: 'closedBy',
    keyDatetime: 'datetime',
    keyEnvId: 'envId',
    keyBundleId: 'bundleId',
    keyNotificationType: 'notificationType',
    keyResult: 'result',
    keyInstanceId: 'instanceId',
    keyInputCount: 'inputCount',
    keyUpdateCount: 'updateCount',
    keyTargetOperation: 'targetOperation',
    keyNlsMessageType: 'nlsMessageType',
    keyEndpointSourceType: 'endpointSourceType',
    keyEndpointTargetType: 'endpointTargetType',
    keyEndpointSourceName: 'endpointSourceName',
    keyEndpointTargetName: 'endpointTargetName',
    keyDebugUrl: 'debugUrl',
    keyNodeID: 'nodeID',
    keyNodeType: 'nodeType',
    keyMessageId: 'messageId',
    keyTransaction: 'transaction',
    keyTransactionType: 'transactionType',
    keyClient: 'client',
    keyPlCpu: 'plCpu',
    keyPlMem: 'plMem',
    keyPlLoad: 'plLoad',
    keyPlUsedDisk: 'plUsedDisk',
    keyPlGwRxTp: 'plGwRxTp',
    keyPlGwTxTp: 'plGwTxTp',
    keyGwMem: 'gwMem',
    keyGwTranRate: 'gwTranRate',
    keyGwMeanTranTime: 'gwMeanTranTime',
    keyRequestProtocol: 'requestProtocol',
    keyQueryString: 'queryString',
    keyRequestMethod: 'requestMethod',
    keyRequestBody_o: 'requestBody_o',
    keyRequestBody: 'requestBody',
    keyRequestBody_a: 'requestBody_a',
    keyRequestHttpHeaders: 'requestHttpHeaders',
    keyStatusCode: 'statusCode',
    keyResponseHttpHeaders: 'responseHttpHeaders',
    keyResponseBody_o: 'responseBody_o',
    keyResponseBody: 'responseBody',
    keyResponseBody_a: 'responseBody_a',
    keyApiId: 'apiId',
    keyApiName: 'apiName',
    keyEnvName: 'envName',
    keyOrgId: 'orgId',
    keyOrgName: 'orgName',
    keyApimAppId: 'apimAppId',
    keyApimAppName: 'apimAppName',
    keyApiVersion: 'apiVersion',
    keyDevOrgId: 'devOrgId',
    keyDevOrgName: 'devOrgName',
    keyPlanId: 'planId',
    keyPlanName: 'planName',
    keyResourceId: 'resourceId',
    keyResourceName: 'resourceName',
    keyPlanVersion: 'planVersion',
    keyDebug: 'debug',
    keyRateLimit: 'rateLimit',
    keyLogPolicy: 'logPolicy',
    keyUserID: 'userID',
    keyAppEvents: 'appEvents',
    keyAppSessionID: 'appSessionID',
    keyCount : 'count',
    keyCrashMessage: 'crashMessage',
    //Miscellaneous
    FORWARDSLASH: '/',
    COLON: ':',
    HTTP: 'http',
    HTTPS: 'https',

    // Internal events which are emitted from SDK if process.env.TESTONLY_EMIT_INTERNAL_EVENTS is set.
    internal_event_load_types_done: 'load_types_done',
    internal_event_load_types_error: 'load_types_error'

};
